petsc: refactor using mpi and p4est

cburstedde opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

The petsc library for scientific computation can make use of the p4est library.
Now that p4est is in nixpkgs, I've added p4est and mpi as dependencies.
Tested with p4est on and off and both openmpi and mpich on and off.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

cburstedde committed:
petsc: refactor using mpi and p4est

cburstedde commented:
This may be of interest to current petsc maintainer @wucke13.

wucke13 commented:
@cburstedde awesome to have some senior academics on board :) Are you aware of `nixfmt`? I'm kind of torn here, on the one hand I'd like to have everything formatted canonical (using `nixfmt`, for example), on the other hand I totally get why you would want to organize the arguments attributes each in a new line... Keep the formatting at your choice, just be aware of the tool if it every comes handy :smile: 

__But__: This seems to break the build of `getdp` for me. I'm not yet entirely sure why, yet. My tip, you could use `nix-review pr 128856` to spot these kind of errors - IIRC it build all packages which depend on the changed one. If you are already using flakes, that would be `nix run nixpkgs#nixpkgs-review pr 128856`.

I really appreciate that you volunteer to be a maintainer for `petsc`, but we have to fix the build error on `getdp`.

r-rmcgibbo commented:
Result of `nixpkgs-review pr 128856` at 72c23a34 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/2fa4b4dc62cd00130ccc3ce175d6a291">getdp</a></li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

markuskowa commented on pkgs/development/libraries/science/math/petsc/default.nix:
```diff
@@ -9,7 +27,12 @@ stdenv.mkDerivation rec {
+  nativeBuildInputs = [ python gfortran gfortran.cc.lib ];
+  buildInputs = [ blas lapack ]
+    ++ lib.optional mpiSupport mpi
+    ++ lib.optional (mpiSupport && mpi == openmpi) openssh
```
```suggestion
    ++ lib.optional (mpiSupport && mpi.pname == "openmpi") openssh
```
Your version compares paths, which could potentially be different if `mpi` and `openmpi` are different variants of `openmpi` (e.g. due to an overlay).

wucke13 replied:
Very good catch!

markuskowa commented on pkgs/development/libraries/science/math/petsc/default.nix:
```diff
@@ -1,5 +1,23 @@
+, blas
+, lapack
+, mpi                   # generic mpi dependency
+, openmpi               # to compare against mpi
```
```suggestion
```

markuskowa commented on pkgs/development/libraries/science/math/petsc/default.nix:
```diff
@@ -30,23 +53,35 @@ stdenv.mkDerivation rec {
-      "--with-fc=$FC"
-      "--with-mpi=0"
+      ${if !mpiSupport then ''
+        "--CC=$CC"
```
Is it necessary to set `CC` etc. in the non-mpi case?

cburstedde replied:
Seems to be working! The --with-mpi=0 is required though.

jedbrown commented on pkgs/development/libraries/science/math/petsc/default.nix:
```diff
@@ -30,23 +53,35 @@ stdenv.mkDerivation rec {
-      Library of linear algebra algorithms for solving partial differential
-      equations
-    '';
+    description = "Linear algebra algorithms for solving partial differential equations";
```
```suggestion
    description = "Parallel solvers for differential algebraic equations and optimization";
```

jedbrown replied:
or
```suggestion
    description = "Portable Extensible Toolkit for Scientific computation";
```

jedbrown commented on pkgs/development/libraries/science/math/petsc/default.nix:
```diff
@@ -30,23 +53,35 @@ stdenv.mkDerivation rec {
+      ${if !mpiSupport then ''
+        "--CC=$CC"
+        "--with-cxx=$CXX"
+        "--with-fc=$FC"
```
Should there be a non-Fortran build? (I don't know nixpkgs priorities; many users don't need Fortran.)

cburstedde committed:
petsc: address couple suggestions

cburstedde committed:
petsc: simplify blas, lapack dependencies

cburstedde committed:
getdp: align with latest petsc dependency

cburstedde committed:
petsc: assert requirement for p4est.mpiSupport

cburstedde committed:
petsc: remove unnecessary configure options

markuskowa commented:
@GrahamcOfBorg eval

cburstedde commented:
Thanks for commenting and merging all.

> Should there be a non-Fortran build? (I don't know nixpkgs priorities; many users don't need Fortran.)

I wouldn't know; it may get a bit too combinatoric doing that eventually.

One question I have: When petsc configures with p4est, it assumes an mpi build.
This breaks when both petsc and p4est are intended to be built without mpi.  We
should at this point have a namespace-clean MPI replacement in p4est, which
might make this possible without too much to change in petsc.

cburstedde commented:
> @cburstedde awesome to have some senior academics on board :) Are you aware of `nixfmt`? I'm kind of torn here, on the one hand I'd like to have everything formatted canonical (using `nixfmt`, for example), on the other hand I totally get why you would want to organize the arguments attributes each in a new line... Keep the formatting at your choice, just be aware of the tool if it every comes handy smile

I went with my understanding of the guidelines, may have missed out on some. Feel free to realign if you like.

> **But**: This seems to break the build of `getdp` for me. I'm not yet entirely sure why, yet. My tip, you could use `nix-review pr 128856` to spot these kind of errors - IIRC it build all packages which depend on the changed one. If you are already using flakes, that would be `nix run nixpkgs#nixpkgs-review pr 128856`.
> 
> I really appreciate that you volunteer to be a maintainer for `petsc`, but we have to fix the build error on `getdp`.

Thanks for your kind welcome. Let me know if there's anything left to address.

cburstedde commented on pkgs/development/libraries/science/math/petsc/default.nix:
```diff
@@ -30,23 +53,35 @@ stdenv.mkDerivation rec {
+        "--with-zlib-lib=-L${zlib}/lib -lz"
+      '' else ""}
       "--with-blas-lib=[${blas}/lib/libblas.so,${gfortran.cc.lib}/lib/libgfortran.a]"
       "--with-lapack-lib=[${lapack}/lib/liblapack.so,${gfortran.cc.lib}/lib/libgfortran.a]"
```
Let's hope the current, generic form works with other implementations.

cburstedde replied:
Just tested with blis 32, builds and checks ok on my home box.

markuskowa commented:
Result of `nixpkgs-review pr 128856` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>getdp</li>
    <li>petsc</li>
  </ul>
</details>

markuskowa commented:
> One question I have: When petsc configures with p4est, it assumes an mpi build.
This breaks when both petsc and p4est are intended to be built without mpi.  We
should at this point have a namespace-clean MPI replacement in p4est, which
might make this possible without too much to change in petsc.

@cburstedde You could change the `assert` to make sure that `p4est` and `petsc` both have idential `mpiSupport` flags. That would solve the problem?

cburstedde commented:
> > One question I have: When petsc configures with p4est, it assumes an mpi build.
> This breaks when both petsc and p4est are intended to be built without mpi.  We
> should at this point have a namespace-clean MPI replacement in p4est, which
> might make this possible without too much to change in petsc.
> 
> @cburstedde You could change the `assert` to make sure that `p4est` and `petsc` both have idential `mpiSupport` flags. That would solve the problem?

The build already guarantees that petsc and p4est have the same mpi status
whenever p4est is active.  The assertion, otoh, points to issues outside nix.

Afaict, the root cause of the issue is petsc configure requesting with-mpi=1
whenever with-p4est is activated.  This is precisely the situation I'm
asserting -- I'd propose to leave it at that until this case is resolved with
future versions of the software packages.

jedbrown commented:
FWIW, `--with-mpi=1` is the default (PETSc configure looks for an MPI if not explicitly specified; nixpkgs will have it in a unique location and thus needs to specify using `--with-mpi-dir=...` or similar).

Having PETSc support p4est without MPI would involve
1. `if self.mpi.found:` before `--enable-mpi` https://gitlab.com/petsc/petsc/-/blob/main/config/BuildSystem/config/packages/p4est.py#L34
2. Ensuring that an `MPI_Comm` is not passed to p4est from C code (compiler should tell you in an attempted build).

cburstedde commented:
> FWIW, `--with-mpi=1` is the default (PETSc configure looks for an MPI if not explicitly specified; nixpkgs will have it in a unique location and thus needs to specify using `--with-mpi-dir=...` or similar).

That matches with my experience.  We can leave the nix logic as is since petsc
builds fine with and without mpi (tested without mpi and p4est both).

> Having PETSc support p4est without MPI would involve
> 1. `if self.mpi.found:` before `--enable-mpi` https://gitlab.com/petsc/petsc/-/blob/main/config/BuildSystem/config/packages/p4est.py#L34
> 2. Ensuring that an `MPI_Comm` is not passed to p4est from C code (compiler should tell you in an attempted build).

I'm too constrained to look into 1. myself.  For 2., it would be ok to pass
sc_MPI_COMM_WORLD to p4est without MPI, which gets replaced trivially.

wucke13 reviewed (approved):
LGTM

markuskowa commented:
`petsc` fails on `aarch64-linux`. Can you please take a look if the build failure the is fixable. If not please just disable it on `aarch64-linux`.

cburstedde committed:
petsc: do not build with p4est on aarch64

cburstedde commented:
> `petsc` fails on `aarch64-linux`. Can you please take a look if the build failure the is fixable. If not please just disable it on `aarch64-linux`.

I have a pending PR on p4est to work around the aarch64 build issue. In the meantime, disabling p4est support on aarch64.

Since you know about openmpi on aarch64: is there anything special required to `make check -j` when the Makefile calls `mpiexec -n 2` on a bunch of mpi executables just built? On linux I must use -j1 to prevent the bot from running out of runners (on my machine it works ok). On aarch64 even with -j1 all tests fail (don't start up properly).

cburstedde committed:
petsc: do not build with p4est on darwin

SuperSandro2000 commented on pkgs/development/libraries/science/math/petsc/default.nix:
```diff
@@ -30,23 +54,31 @@ stdenv.mkDerivation rec {
     patchShebangs .
```
Can we make this more precise?

cburstedde replied:
Can you help me out please -- what do you mean?

The blurb on the `hot shebang stuff` predates my involvement, and I don't know the petsc build system.

SuperSandro2000 replied:
> Can you help me out please -- what do you mean?

Instead of trying to patch every file it would be better to only specify the file or directory that really needs patching. This would be a lot faster.

cburstedde replied:
Looks like: src systems/Apple lib/petsc/bin setup.py config

@jedbrown would that be missing anything now or in the future?

cburstedde replied:
Which is for --host and which for --build I wouldn't want to guess, so maybe just using the paths is ok.

jedbrown replied:
The build system itself generally uses `$(PYTHON)` and the like (as set in configure) rather than relying on shebangs. There are some installed utilities in `lib/petsc/bin/` that would need shebangs.

jedbrown replied:
You can run `/path/to/my/python configure --options` instead of hacking the script.

SuperSandro2000 replied:
> Which is for --host and which for --build I wouldn't want to guess, so maybe just using the paths is ok.

You don't need to specify this. Just list the paths you want to patch. Having . is very slow for a lot of files and should be avoided.

cburstedde replied:
Tried this and it's working. I'm flying blind here, no guarantees on purity etc.

I found that `lib/petsc/bin/petscmpiexec` uses `csh`, which is not in nix currently.
There are also installed files that contain `/usr/bin/env python` on the first line:

 * result/lib/petsc/bin/configureTAS.py
 * result/lib/petsc/bin/extract.py

cburstedde committed:
Merge branch 'master' into package-petsc-p4est

markuskowa commented:
> Since you know about openmpi on aarch64: is there anything special required to make check -j when the Makefile calls mpiexec -n 2 on a bunch of mpi executables just built? On linux I must use -j1 to prevent the bot from running out of runners (on my machine it works ok). On aarch64 even with -j1 all tests fail (don't start up properly).

I am not sure what exactly goes wrong on aarch64 but in connection with MPI there are a few standard pitfalls/re-appearing patterns:
* `-n 2` will fail if `openmpi` detects only 1 CPU. Setting `OMPI_MCA_rmaps_base_oversubscribe=1` in the `preCheck` phase helps here.
* For everything that is not openmpi but MPICH based one needs to set `HYDRA_IFACE=lo` or the build may fail in a sandbox.
* Not directly MPI related, but setting `OMP_NUM_THREADS=N` to reasonable value helps to avoid an overloaded build machine (happens usually on big machines with a lot of CPUs).

cburstedde commented:
> > Since you know about openmpi on aarch64: is there anything special required to make check -j when the Makefile calls mpiexec -n 2 on a bunch of mpi executables just built? On linux I must use -j1 to prevent the bot from running out of runners (on my machine it works ok). On aarch64 even with -j1 all tests fail (don't start up properly).
> 
> I am not sure what exactly goes wrong on aarch64 but in connection with MPI there are a few standard pitfalls/re-appearing patterns:
> 
>     * `-n 2` will fail if `openmpi` detects only 1 CPU. Setting `OMPI_MCA_rmaps_base_oversubscribe=1` in the `preCheck` phase helps here.

Thanks, will try!

>     * For everything that is not openmpi but MPICH based one needs to set `HYDRA_IFACE=lo` or the build may fail in a sandbox.

So this also applies to preCheck, as in `export HYDRA_IFACE=lo`?

>     * Not directly MPI related, but setting `OMP_NUM_THREADS=N` to reasonable value helps to avoid an overloaded build machine (happens usually on big machines with a lot of CPUs).

I'm not compiling in OpenMP currently, hoping to get away without setting anything (just in case: `N = np` or `N = 1`?).

markuskowa commented:
This is a typical template I use for MPI/OpenMP packages which have a check phase:
```
preCheck = ''
  export OMP_NUM_THREADS=1 # or 2 if you want to also test some OpenMP parallelism.
  export HYDRA_IFACE=lo
  export OMPI_MCA_rmaps_base_oversubscribe=1
'';
```

cburstedde committed:
petsc: restrict patchSheBang; use configureScript

The configure script can be called with python via configureScript.
This leaves only ./lib/petsc/bin to be patched.

cburstedde committed:
Merge branch 'master' into package-petsc-p4est

cburstedde committed:
petsc: build with p4est for all platforms

Recent master updates make p4est build on aarch64 and darwin.
Undo their deactivation in the petsc build.

SuperSandro2000 commented on pkgs/development/libraries/science/math/petsc/default.nix:
```diff
@@ -27,26 +51,35 @@ stdenv.mkDerivation rec {
+      "--with-lapack=1"
     )
   '';
+  configureScript = "python ./configure";
```
```suggestion
  preConfigure = ''
    python ./configure
  '';
```
Does that work, too?

cburstedde replied:
But then configure would run twice, once from preConfigure (with python) and once from Configure (with bash)?

Imho the current form looks exactly like petsc intends it to be used.

SuperSandro2000 replied:
After checking the docs we can leave it like this.

SuperSandro2000 commented on pkgs/development/libraries/science/math/petsc/default.nix:
```diff
@@ -1,4 +1,20 @@
+, fetchurl
+, darwin
+, gfortran
+, python
```
Can this be python3? Python2 should no longer be used in new things.

jedbrown replied:
Yeah, we left it that way because the scripts work in both Python2 and Python3, and some ancient HPC systems with large user bases still have Python2. `python` is Python3 on some systems (since ~2011 on Arch).

cburstedde replied:
What is preferred then, hardcoding `python3` here or leaving it at `python`, which can be overridden system wide as needed (myself favoring the latter, but I'm no python expert).

SuperSandro2000 replied:
Python2 is EOL and should no longer be used in new packages. Just use python3.

> leaving it at python, which can be overridden system wide as needed (myself favoring the latter, but I'm no python expert).

I am not sure if this is something nixpkgs officially supports.

cburstedde committed:
petsc: specify python3 and remove unneeded patch
